Cabal/src/Distribution/Simple/Program/Run.hs

```diff
@@ -248,6 +249,17 @@ getEffectiveEnvironment overrides =
     update (var, Nothing) = Map.delete var
     update (var, Just val) = Map.insert var val
 
+-- | Like 'getEffectiveEnvironment', but when no overrides are specified,
+-- returns the full environment instead of 'Nothing'.
+getFullEnvironment
+  :: [(String, Maybe String)]
+  -> IO [(String, String)]
+getFullEnvironment overrides = do
+  menv <- getEffectiveEnvironment overrides
+  case menv of
+    Just env -> return env
+    Nothing -> getEnvironment
```

Comment:
I don't like this code, because it relies on the following invariant of `getEffectiveEnvironment`: it only returns `Nothing` when passed an empty list. Without knowing this, it looks that `getEffectiveEnvironment` can "fail" and that in that case we are discarding the overrides. I suppose it makes sense if you think of `getEffectiveEnvironment` as returning exactly what you would put in the `env` field of `CallProcess` (where `Nothing` means: use the ambient environment).

If you don't want to change any of that, could you at least add a comment, e.g. "`getEffectiveEnvironment` returns `Nothing` precisely when there are no overrides"?

Reply:
I guess by "add a comment" @sheaf meant a comment in the implementation. @mpickering do you want to do that?